File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/FilesPage/Export/TranslationExchangeExport.java
package org.wycliffeassociates.translationrecorder.FilesPage.Export;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.preference.PreferenceManager;
import android.util.Log;
import com.door43.tools.reporting.Logger;
import net.gotev.uploadservice.*;
import org.wycliffeassociates.translationrecorder.FilesPage.FeedbackDialog;
import org.wycliffeassociates.translationrecorder.R;
import org.wycliffeassociates.translationrecorder.SettingsPage.Settings;
import org.wycliffeassociates.translationrecorder.TranslationRecorderApp;
import org.wycliffeassociates.translationrecorder.database.ProjectDatabaseHelper;
import org.wycliffeassociates.translationrecorder.project.Project;
import org.wycliffeassociates.translationrecorder.project.components.User;

import java.io.File;
import java.util.UUID;

/**
 * Created by sarabiaj on 11/16/2017.
 */

public class TranslationExchangeExport extends Export implements UploadStatusDelegate {

    public static int EXPORT_UPLOAD_ID = 3;

    TranslationExchangeDiff mDiffer;
    ProjectDatabaseHelper db;
    UploadServiceSingleBroadcastReceiver uploadReceiver;

    public TranslationExchangeExport(File projectToExport, Project project, ProjectDatabaseHelper db) {
        super(projectToExport, project);
        mDirectoryToZip = null;
        this.db = db;
        uploadReceiver = new UploadServiceSingleBroadcastReceiver(this);
    }

    @Override
    protected void initialize() {
        mDiffer = new TranslationExchangeDiff(
                (TranslationRecorderApp) mCtx.getActivity().getApplication(),
                mProject
        );
        mDiffer.computeDiff(outputFile(), this);
    }

    @Override
    public void onStart(final int id) {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                if (id == TranslationExchangeDiff.DIFF_ID) {
                    mProgressCallback.setProgressTitle("Step 1/3: Generating manifest file");
                } else if(id == ZipProject.ZIP_PROJECT_ID) {
                    mProgressCallback.setProgressTitle("Step 2/3: Packaging files to export");
                } else if(id == TranslationExchangeExport.EXPORT_UPLOAD_ID) {
                    mProgressCallback.setProgressTitle("Step 3/3: Uploading");
                }

                mZipDone = false;
                mProgressCallback.setZipping(true);
                mProgressCallback.showProgress(ProgressUpdateCallback.ZIP);
            }
        });
    }

    @Override
    public void onComplete(int id) {
        super.onComplete(id);
        if(id == TranslationExchangeDiff.DIFF_ID) {
            mFilesToZip = mDiffer.getDiff();
            super.initialize();
        } else if(id == EXPORT_UPLOAD_ID) {
            mProgressCallback.setProgressTitle(null);
            mProgressCallback.setZipping(false);
        }
    }

    @Override
    protected void handleUserInput() {
        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                Context ctx = mCtx.getActivity().getApplicationContext();
                uploadBinary(ctx, outputFile());
            }
        });
        thread.start();
    }

    public void uploadBinary(Context context, File file) {
        try {
            this.onStart(EXPORT_UPLOAD_ID);

            // starting from 3.1+, you can also use content:// URI string instead of absolute file
            String filePath = file.getAbsolutePath();
            SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
            int userId = pref.getInt(Settings.KEY_USER, 1);
            User user = db.getUser(userId);
            String hash = user.getHash();

            String uploadId = UUID.randomUUID().toString();
            uploadReceiver.setUploadID(uploadId);
            uploadReceiver.register(context);

            new BinaryUploadRequest(context, uploadId, "http://opentranslationtools.org/api/upload/zip")
                            .addHeader("tr-user-hash", hash)
                            .addHeader("tr-file-name", file.getName())
                            .setFileToUpload(filePath)
                            .setNotificationConfig(getNotificationConfig())
                            .setDelegate(null)
                            .setAutoDeleteFilesAfterSuccessfulUpload(true)
                            .setMaxRetries(30)
                            .startUpload();

        } catch (Exception exc) {
            Log.e("AndroidUploadService", exc.getMessage(), exc);
        }
    }

    protected UploadNotificationConfig getNotificationConfig() {
        UploadNotificationConfig config = new UploadNotificationConfig();

        config.getProgress().iconResourceID = R.drawable.ic_upload;
        config.getProgress().iconColorResourceID = Color.BLUE;

        config.getCompleted().iconResourceID = R.drawable.ic_upload_success;
        config.getCompleted().iconColorResourceID = Color.GREEN;

        config.getError().iconResourceID = R.drawable.ic_upload_error;
        config.getError().iconColorResourceID = Color.RED;

        config.getCancelled().iconResourceID = R.drawable.ic_cancelled;
        config.getCancelled().iconColorResourceID = Color.YELLOW;

        return config;
    }

    @Override
    public void onProgress(Context context, UploadInfo uploadInfo) {
        this.setUploadProgress(EXPORT_UPLOAD_ID, uploadInfo.getProgressPercent());
    }

    @Override
    public void onCompleted(Context context, UploadInfo uploadInfo, ServerResponse serverResponse) {
        FeedbackDialog fd = FeedbackDialog.newInstance(
                "Project upload",
                "Project has been successfully uploaded."
        );
        fd.show(mCtx.getFragmentManager(), "title");

        mZipFile.delete();

        Logger.e(
                TranslationExchangeExport.class.toString(),
                "code: " + serverResponse.getHttpCode() + " " + serverResponse.getBodyAsString()
        );
        this.onComplete(EXPORT_UPLOAD_ID);
    }

    @Override
    public void onError(Context context, UploadInfo uploadInfo, ServerResponse serverResponse, Exception exception) {
        String message;
        if (serverResponse != null) {
            message = String.format("code: %s: %s",
                    serverResponse.getHttpCode(),
                    serverResponse.getBodyAsString()
            );
            Logger.e(TranslationExchangeExport.class.toString(), message, exception);
        } else if (exception != null) {
            message = exception.getMessage();
            Logger.e(TranslationExchangeExport.class.toString(), "Error: " + message, exception);
        } else {
            message = "An error occurred without a response or exception, upload percent is "
                    + uploadInfo.getProgressPercent();
            Logger.e(TranslationExchangeExport.class.toString(), message);
        }

        FeedbackDialog fd = FeedbackDialog.newInstance(
                "Project upload",
                "Project upload failed: " + message
        );
        fd.show(mCtx.getFragmentManager(), "UPLOAD_FEEDBACK");
        this.onComplete(EXPORT_UPLOAD_ID);
    }

    @Override
    public void onCancelled(Context context, UploadInfo uploadInfo) {
        Logger.e(TranslationExchangeExport.class.toString(), "Cancelled upload");
        if (uploadInfo != null) {
            Logger.e(
                    TranslationExchangeExport.class.toString(),
                    "Upload percent was " + uploadInfo.getProgressPercent()
            );
        }
        this.onComplete(EXPORT_UPLOAD_ID);
    }
}

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/FilesPage/Export/TranslationExchangeExport.kt
package org.wycliffeassociates.translationrecorder.FilesPage.Export

import android.content.Context
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.door43.tools.reporting.Logger
import net.gotev.uploadservice.data.UploadInfo
import net.gotev.uploadservice.data.UploadNotificationConfig
import net.gotev.uploadservice.data.UploadNotificationStatusConfig
import net.gotev.uploadservice.exceptions.UploadError
import net.gotev.uploadservice.exceptions.UserCancelledUploadException
import net.gotev.uploadservice.network.ServerResponse
import net.gotev.uploadservice.observer.request.RequestObserverDelegate
import net.gotev.uploadservice.protocols.binary.BinaryUploadRequest
import org.wycliffeassociates.translationrecorder.FilesPage.FeedbackDialog
import org.wycliffeassociates.translationrecorder.R
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity
import org.wycliffeassociates.translationrecorder.TranslationRecorderApp
import org.wycliffeassociates.translationrecorder.database.IProjectDatabaseHelper
import org.wycliffeassociates.translationrecorder.persistance.AssetsProvider
import org.wycliffeassociates.translationrecorder.persistance.IDirectoryProvider
import org.wycliffeassociates.translationrecorder.persistance.IPreferenceRepository
import org.wycliffeassociates.translationrecorder.persistance.getDefaultPref
import org.wycliffeassociates.translationrecorder.project.Project
import java.io.File

/**
 * Created by sarabiaj on 11/16/2017.
 */
class TranslationExchangeExport(
    project: Project,
    private val db: IProjectDatabaseHelper,
    private val directoryProvider: IDirectoryProvider,
    private val prefs: IPreferenceRepository,
    private val assetsProvider: AssetsProvider,
) : Export(project, directoryProvider), RequestObserverDelegate {

    private var differ: TranslationExchangeDiff? = null

    init {
        directoryToZip = null
    }

    override fun initialize() {
        differ = TranslationExchangeDiff(
            project,
            db,
            directoryProvider,
            assetsProvider,
            prefs
        ).apply {
            outputFile()
            computeDiff(this@TranslationExchangeExport)
        }
    }

    override fun onStart(id: Int) {
        handler.post {
            when (id) {
                TranslationExchangeDiff.DIFF_ID -> {
                    progressCallback?.setProgressTitle(fragment.getString(R.string.upload_step_one))
                }
                ZipProject.ZIP_PROJECT_ID -> {
                    progressCallback?.setProgressTitle(fragment.getString(R.string.upload_step_two))
                }
                EXPORT_UPLOAD_ID -> {
                    progressCallback?.setProgressTitle(fragment.getString(R.string.upload_step_three))
                }
            }

            zipDone = false
            progressCallback?.setZipping(true)
            progressCallback?.showProgress(ProgressUpdateCallback.ZIP)
        }
    }

    override fun onComplete(id: Int) {
        super.onComplete(id)
        if (id == TranslationExchangeDiff.DIFF_ID) {
            filesToZip = differ?.diff
            super.initialize()
        } else if (id == EXPORT_UPLOAD_ID) {
            progressCallback?.setProgressTitle(null)
            progressCallback?.setZipping(false)
        }
    }

    override fun handleUserInput() {
        val thread = Thread {
            val ctx = fragment.requireActivity().applicationContext
            uploadBinary(ctx, outputFile())
        }
        thread.start()
    }

    private fun uploadBinary(context: Context, file: File) {
        try {
            this.onStart(EXPORT_UPLOAD_ID)

            val uploadServer = prefs.getDefaultPref(
                SettingsActivity.KEY_PREF_UPLOAD_SERVER,
                "http://opentranslationtools.org"
            )

            // starting from 3.1+, you can also use content:// URI string instead of absolute file
            val filePath = file.absolutePath
            val userId = prefs.getDefaultPref(SettingsActivity.KEY_PROFILE, 1)
            val user = db.getUser(userId)!!
            val hash = user.hash!!

            val request = BinaryUploadRequest(context, "$uploadServer/api/upload/zip")
                .addHeader("tr-user-hash", hash)
                .addHeader("tr-file-name", file.name)
                .setFileToUpload(filePath)
                .setNotificationConfig { _, _ ->  notificationConfig}
                .setAutoDeleteFilesAfterSuccessfulUpload(true)
                .setMaxRetries(30)

            Handler(Looper.getMainLooper()).post {
                request.subscribe(fragment.requireContext(), fragment, this)
            }
        } catch (exc: Exception) {
            Log.e("AndroidUploadService", exc.message, exc)
        }
    }

    private val notificationConfig: UploadNotificationConfig
        get() {
            val progressConfig = UploadNotificationStatusConfig(
                "Progress",
                "Upload progress.",
                R.drawable.ic_upload,
                Color.BLUE
            )
            val successConfig = UploadNotificationStatusConfig(
                "Success",
                "Successfully uploaded.",
                R.drawable.ic_upload_success,
                Color.GREEN
            )
            val errorConfig = UploadNotificationStatusConfig(
                "Error",
                "An error occurred.",
                R.drawable.ic_upload_error,
                Color.RED
            )
            val cancelConfig = UploadNotificationStatusConfig(
                "Canceled",
                "Canceled by the user.",
                R.drawable.ic_cancelled,
                Color.YELLOW
            )

            val config = UploadNotificationConfig(
                TranslationRecorderApp.NOTIFICATION_CHANNEL_ID,
                true,
                progressConfig,
                successConfig,
                errorConfig,
                cancelConfig
            )
            return config
        }

    override fun onProgress(context: Context, uploadInfo: UploadInfo) {
        this.setUploadProgress(EXPORT_UPLOAD_ID, uploadInfo.progressPercent)
    }

    override fun onSuccess(
        context: Context,
        uploadInfo: UploadInfo,
        serverResponse: ServerResponse
    ) {
        val fd = FeedbackDialog.newInstance(
            this.fragment.getString(R.string.project_upload),
            this.fragment.getString(R.string.project_uploaded)
        )
        fd.show(this.fragment.parentFragmentManager, "title")

        zipFile?.delete()

        Logger.e(
            TranslationExchangeExport::class.java.toString(),
            "code: " + serverResponse.code + " " + serverResponse.bodyString
        )
        this.onComplete(EXPORT_UPLOAD_ID)
    }

    override fun onCompleted(context: Context, uploadInfo: UploadInfo) {
    }

    override fun onCompletedWhileNotObserving() {
    }

    override fun onError(context: Context, uploadInfo: UploadInfo, exception: Throwable) {
        val message: String
        when (exception) {
            is UserCancelledUploadException -> {
                Logger.e(TranslationExchangeExport::class.java.toString(), "Cancelled upload")
                Logger.e(
                    TranslationExchangeExport::class.java.toString(),
                    "Upload percent was " + uploadInfo.progressPercent
                )
                message = "Upload was cancelled by user: ${exception.message ?: ""}"
                this.onComplete(EXPORT_UPLOAD_ID)
            }
            is UploadError -> {
                message = String.format(
                    "code: %s: %s",
                    exception.serverResponse.code,
                    exception.serverResponse.bodyString
                )
                Logger.e(TranslationExchangeExport::class.java.toString(), message, exception)
                Logger.e(
                    TranslationExchangeExport::class.java.toString(),
                    "Error: $exception.message", exception
                )
            }
            else -> {
                Logger.e(
                    TranslationExchangeExport::class.java.toString(),
                    "Error: $uploadInfo", exception
                )
                message = "Error: ${exception.message ?: ""}"
            }
        }

        val fd = FeedbackDialog.newInstance(
            this.fragment.getString(R.string.project_upload),
            this.fragment.getString(R.string.project_upload_failed, message)
        )
        fd.show(this.fragment.parentFragmentManager, "UPLOAD_FEEDBACK")
        this.onComplete(EXPORT_UPLOAD_ID)
    }

    companion object {
        var EXPORT_UPLOAD_ID: Int = 3
    }
}
